### ISSUE TITLE
App doesn't update risk status (in foreground)

### ISSUE THREAD

**Issue by Ein-Tim** [corona-warn-app/cwa-app-ios#1884]
## Avoid duplicates

* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is not already reported in another issue

## Technical details

- Device name: iPhone XR
- iOS Version: 14.4
- App Version: 1.11.0

- Case 2 @cricketsamya:  iPhone 12 Pro Max /CWA 1.11.0 (9)/iOS 14.4

## Describe the bug

I opened CWA at 15:40, more than 4 hours have passed since the last check (11:27). 
The app did NOT automatically carry out a risk check. 
Restarting the app did not help, but deactivating background updates did. 
Seems like a problem with automatic updates.

## Steps to reproduce the issue

1. Open App after 4 hours have passed since the last check
2. See that no check happens 
3. Deactivate background updates in iOS Settings
4. See that the check instantly happens when you reopen the app

## Expected behaviour

A risk check should, if the phone is connected to WiFi, happen after 4 hours.

## Additional context

No ENA entries in settings. 

EN-Log: 
[ExposureChecks-2021-01-29.txt](https://github.com/corona-warn-app/cwa-app-ios/files/5894458/ExposureChecks-2021-01-29.txt)

<details>
<summary>Screenrecordings:</summary>

https://user-images.githubusercontent.com/67682506/106321370-74308f80-6274-11eb-841b-d9352d9187da.MOV

https://user-images.githubusercontent.com/67682506/106356004-6888ab80-62fc-11eb-81b0-0ae5ddcea2d1.mp4

</details>

Could not reproduce this on an iPhone 6s. Will continue monitoring.

The check often happens shortly after the app was closed, in the background.

This issue persists after restarting the device. 

---
Internal tracking ID: [EXPOSUREAPP-4953](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4953)

**Comment by cricketsamya** [comment]
![image](https://user-images.githubusercontent.com/13750637/106566848-2a280200-6531-11eb-85bf-2f3b5efbcfd7.png)

App is not updating Risk status since yesterday 01.02.2021

**Comment by Ein-Tim** [comment]
@cricketsamya Are you using iOS 14.4?

Interesting is that your app is also not updating in the background, mine is only not updating in the foreground. 

Did you restart your device?

**Comment by dsarkar** [comment]
iPhone6s / iOS 14.4 / CWA 1.11 - cannot reporduce - I will continue observing issue.
@Ein-Tim FYI
@cricketsamya iPhone modell ? / CWA 1.11 / ios 14.4 ? Thanks for feedback.

**Comment by dsarkar** [comment]
@Ein-Tim 
- You reported 4 days ago this issue. Any pattern you can report. 
- Can you reproduce the issue on the iPhone XR? If so, what steps?
- Can you reproduce the issue on the iPhone 6s? If so, what steps?
Thanks for reporting and the feedback. We are checking also for reports on the AppStore.

**Comment by cricketsamya** [comment]
> @cricketsamya Are you using iOS 14.4?
> 
> Interesting is that your app is also not updating in the background, mine is only not updating in the foreground.
> 
> Did you restart your device?

I am using iOS 14.4. After restart, the app synced. May be it was stuck during last sync?

**Comment by cricketsamya** [comment]
> iPhone6s / iOS 14.4 / CWA 1.11 - cannot reproduce - I will continue observing issue.
> @Ein-Tim FYI
> @cricketsamya iPhone modell ? / CWA 1.11 / ios 14.4 ? Thanks for feedback.

@dsarkar 
iPhone 12 Pro Max /CWA 1.11.0 (9)/iOS 14.4

**Comment by Ein-Tim** [comment]
@cricketsamya 

>I am using iOS 14.4. After restart, the app synced. May be it was stuck during last sync?

Could you provide your EN-Log here?

<details>
<summary>How to export Exposure Notificaiton Logs under iOS 13.7 or newer</summary>



**German:**

`Einstellungen` ➡️ `Begegnungsmitteilungen` ➡️ `Status von Begegnungsaufzeichnungen` ➡️ `Begegnungsüberprüfungen` ➡️ `Überprüfungen exportieren`

**English:**

`Settings` ➡️ `Exposure Notifications` ➡️ `Exosure Logging Status` ➡️ `Exposure Checks` ➡️ `Export Exposure Checks`

**To upload the file here, rename it from .json to .txt**
</details>

**Comment by Ein-Tim** [comment]
@dsarkar 

>You reported this issue 4 days ago. Any pattern you can report?

Not really, no... The only pattern is that the check always works if I open the app with deactivated background updates.

>Can you reproduce the issue on the iPhone XR? If so, what steps?
>Can you reproduce the issue on the iPhone 6s? If so, what steps?

There are no special steps I do to reproduce this bug. Sometimes it works on the iPhone XR, sometimes not. I did not yet see this on the iPhone 6s.

**Comment by cricketsamya** [comment]
> @cricketsamya
> 
> > I am using iOS 14.4. After restart, the app synced. May be it was stuck during last sync?
> 
> Could you provide your EN-Log here?
> 
> How to export Exposure Notificaiton Logs under iOS 13.7 or newer

@Ein-Tim 
Done! Have a look!
[ExposureChecks-2021-02-02.txt](https://github.com/corona-warn-app/cwa-app-ios/files/5910403/ExposureChecks-2021-02-02.txt)

**Comment by Ein-Tim** [comment]
Thank you @cricketsamya 

When looking at your EN-Log it seems like that the last check (yesterday, 15:53) was not interrupted, but the risk check just stopped working until today at 09:49. 


Update from my side:
Currently I don't see this behaviour anymore, neither on the iPhone XR nor on the iPhone 6s.
I will continue to monitor and try to reproduce it.

**Comment by dsarkar** [comment]
@Ein-Tim @cricketsamya Internal tracking ID: EXPOSUREAPP-4953

**Comment by Ein-Tim** [comment]
I was again able to reproduce this today, @dsarkar is there anything I should do the next time this happens? Sadly there are no ENA files in the analytics files in the settings...

**Comment by dsarkar** [comment]
@Ein-Tim Thanks for the info. I asked the dev he needs more info from your side. Stand by. Fix is planned for 1.12 or 1.13.

**Comment by Ein-Tim** [comment]
@dsarkar 

What information should I provide?

**Comment by dsarkar** [comment]
@Ein-Tim Right now no further information needed from the developer's side, many thanks! There are fixed introduced in 1.12 and 1.13, so monitoring this issue then would be appreciated.

**Comment by Ein-Tim** [comment]
I will continue to monitor with 1.12.
If I don't see this behaviour anymore, I will close this issue.

**Comment by dsarkar** [comment]
@Ein-Tim OK, let´s do that. Thanks.

**Comment by cricketsamya** [comment]
@dsarkar 

iPhone 12 Pro Max
CWA 1.12.1
iOS 14.4
Looks like it appeared after update to CWA 1.12.1
Check exposure logs timestamps.. 
![image](https://user-images.githubusercontent.com/13750637/107886312-477ca900-6eff-11eb-82b5-f8424744ccb9.png)

**Comment by dsarkar** [comment]
@cricketsamya ok, thanks for the screenshot. At first sight, it looks ok to me. 

- If you are on wifi, 4- max 6 downloads per day are typical. 
- On mobile data max one download per day
- If you use mobile data for some hours and then wifi you might have 1 - max 6 downloads per day.

So it looks fine at first sight.

You might try the following: Do not open the app for 24 h and show us then a screenshot. If there us day missing, then it would not look right.

Thanks.

**Comment by cricketsamya** [comment]
@dsarkar I was on mobile data today for few hours. Then wifi all time.. may be that explains the missing syncs.

**Comment by dsarkar** [comment]
- Oh, I see, you are saying that in the morning the first check is only triggered when you open the app? 
- I guess you are using at night wifi, correct?
- In the morning, when the first check happens, wifi or mobile data?
-

**Comment by cricketsamya** [comment]
@dsarkar first check always happens with wifi for sure 💪🏼

**Comment by dsarkar** [comment]
@cricketsamya ok, maybe try this. Don't open the app for a day and report back. Feedback very much appreciated! Thanks

**Comment by cricketsamya** [comment]
@dsarkar I am talking about yesterdays last check! That happens after I opened the app! And todays last check as well..

**Comment by dsarkar** [comment]
@cricketsamya the last check and the check before the last check have a time difference of 4-6 hours in your case, so that looks ok so far. What called my attention, is that during the night there is a larger gap. However, as long you get 4-6 checks per day on wifi, or 1 check on data, or 1-6 check data/wifi mixed, it looks ok so far.

**Comment by Ein-Tim** [comment]
I think it is normal that @cricketsamya does not have any checks at night since he deactivates Bluetooth during the night.

**Comment by cricketsamya** [comment]
@dsarkar @Ein-Tim yes I deactivate bluetooth. I will try the test! And report! Thanks!

**Comment by dsarkar** [comment]
@cricketsamya 
@Ein-Tim ok, thanks for pointing that out! I was not aware of BT off during the night. So, all good. Thanks, and good night.

**Comment by Ein-Tim** [comment]
I, personally, can't reproduce this anymore. I will still leave this open to further investigate @cricketsamya's case. 

@dsarkar What you could maybe ask the devs if it's intended that the App does not update in the foreground if the App is opened in the foreground and 4 hours pass (I hope you understand what you mean).

**Comment by dsarkar** [comment]
@Ein-Tim 

> (I hope you understand what you mean).

Usually I do ;-)

**Comment by Ein-Tim** [comment]
For sure @dsarkar (:

I will try to explain with an example:

Your last check was today at 14:00h, now you open the App at 17:59 and leave it open. You will see that there is no check carried out at exactly 18:00h (what I would expect). The next check happens as soon as you close the app. Now I wonder whether this is intended or not. 

I hope now it's clear what I'm talking about 😁

**Comment by cricketsamya** [comment]
@Ein-Tim @dsarkar 
![image](https://user-images.githubusercontent.com/13750637/107988684-d60d2b00-6fd0-11eb-9f8a-da64212fd1b2.png)

I did not open the app for 24 hours! Looks like it worked! Except risk status still says 12/14.

All good!

**Comment by Ein-Tim** [comment]
@cricketsamya 

Nice to hear that everything is working as it should 👍

I will close this after my question [above](https://github.com/corona-warn-app/cwa-app-ios/issues/1884#issuecomment-779379907) has been answered. 

Have a nice evening!

**Comment by ndegendogo** [comment]
> Your last check was today at 14:00h, now you open the App at 17:59h and leave it open. You will see that there is no check carried out at exactly 18:00h (what I would expect). The next check happens as soon as you close the app. 

@Ein-Tim interesting. How long did you leave the app open? Because also from background I do never have **exact** 4 hours.

Besides that, my expectation *as a user* would be that it should not matter whether I just switch off the display while cwa is still open, or whether I press the home button before. **Of course I expect the check to be triggered after (roughly) every 4 hours in both cases.**

**Comment by Ein-Tim** [comment]
@ndegendogo 

> How long did you leave the app open? 

For at least 3-4 Minutes. But I expect a check to be triggered after **exactly** 4 hours if the app is opened.  

>Because also from background I do never have exact 4 hours

I never have exactly 4 hours in the background, but if I (in the example above) put the app in foreground at 18:00h, a check will happen.

> Besides that, my expectation as a user would be that it should not matter whether I just switch off the display while cwa is still open, or whether I press the home button before. Of course I expect the check to be triggered after (roughly) every 4 hours in both cases.

I think, and I will test this, the check will be triggered after
a) pressing the home button
b) switch of the display
c) killing CWA (unsure about that).

**Comment by dsarkar** [comment]
@Ein-Tim ok, I think I understand now the question you made earlier. In any case, the updates are **not exactly** every 4 hours. Can't find it now, but it was discussed somewhere already.

**Comment by Ein-Tim** [comment]
@dsarkar But I have the stong feeling that the check will not be triggered at all in the foreground in the case I've described above. I will try it and keep you up to date.

**Comment by dsarkar** [comment]
@dsarkar Interesting, I will try it ... need to wait for about 3 h now.

**Comment by ndegendogo** [comment]
@dsarkar in parallel you could try to ask the developers what is the **intended** behaviour.

If I remember correctly, I have seen risk check triggered from start of application and from transition to the home screen; and the timer started from closing the app. This seems inline with @Ein-Tim 's observations. 
*But I may well err, have misunderstood something, overseen something or din't remember well ...*

**Comment by Ein-Tim** [comment]
Tested the following:

(Last check 08:08h)

1. Open App at 12:07h
2. No check happens at 12:08h
3. No check happens between 12:08h - 12:30h if the app is left in foreground
4. Either close the app, open a new view and go back to the home screen or switch off the display -> check happens

<details>
<summary>Screen recordings and screenshots</summary>

12:07h -> 12:08h:

https://user-images.githubusercontent.com/67682506/108064625-3ea1e980-705d-11eb-9d93-908f56d477e0.mp4

Still no risk-check at 12:12h:
<img src="https://user-images.githubusercontent.com/67682506/108064713-5bd6b800-705d-11eb-8d12-060ad810cead.PNG" width="300">


Still no risk-check at 12:30h:
<img src="https://user-images.githubusercontent.com/67682506/108064757-68f3a700-705d-11eb-86e5-38141e055223.PNG" width="300">

Open another view to trigger the check:

https://user-images.githubusercontent.com/67682506/108064790-7446d280-705d-11eb-9a2d-336f4418cbe8.MP4

</details>

**Comment by dsarkar** [comment]
@Ein-Tim Thanks for the feedback. I will also check this over a longer period of time.

**Comment by dsarkar** [comment]
![image](https://user-images.githubusercontent.com/1409741/108710122-e878ee80-7513-11eb-924f-51ab2641feff.png)


@Ein-Tim Last week Tuesday 1600 - Wednesday 1200 we run a check: CWA in the foreground, screen on. See screenshots above, checks are done regularly in the foreground.

**Comment by Ein-Tim** [comment]
@dsarkar Okay, thanks, I consider this as solved now and will close it.

**Comment by dsarkar** [comment]
@Ein-Tim ok! Thanks.

**Comment by ndegendogo** [comment]
@Ein-Tim did you continue to watch this? Any observations?

With 1.12 I had the **feeling** (although no hard data) that the background risk check was more reliable when I returned to the home screen before I pressed the home button to close the app.
When I forgot this the background check seemed unreliable ...

**Comment by Ein-Tim** [comment]
@ndegendogo 

Thanks for the hint, I closed this issue since I didn't experience it anymore. 
I will try it again now on 1.13, if I see this behavior again, I'll reopen the issue.
Did you experience it on 1.13?

**Comment by ndegendogo** [comment]
I don't have loaded 1.13 yet.
And it is a bit hard to test here - during the week I go to the office and have phases without WLAN. Also I am using the contact diary very much, so when I open the app it triggers the check again.
